### ISSUE TITLE
Bump from 1.3.1 to 1.3.2 causes change in decrypted value

### ISSUE THREAD

**Issue by johnmosesman** [rubencaro/cipher#15]
These are dev values so I don't mind sharing them to give you actionable results to try.

```
config :cipher, keyphrase: "just_for_dev_aaaaaaaaaaaaaa",
                ivphrase: "not_important_aaaaaaaaaaaaa"
```

Using 1.3.1, this works:

```
iex(3)> encrypted_value = Repo.get(User, 9).encrypted_api_token
"g7Yd7my6FJOlgRXqCeDDRChoIorKcqXN5qcC0PXei4Ai6VXHYxYFVuevpon%2BIZeQ"
iex(4)> encrypted_value |> Cipher.decrypt
"WMo1D7ccQa9utLhbTw2obnb8ktb7hb4RIe2VW9tNzzM="
iex(5)> encrypted_value |> Cipher.decrypt |> Cipher.encrypt
"g7Yd7my6FJOlgRXqCeDDRChoIorKcqXN5qcC0PXei4Ai6VXHYxYFVuevpon%2BIZeQ"
iex(6)> encrypted_value == (encrypted_value |> Cipher.decrypt |> Cipher.encrypt)
true
```

Using 1.3.2 the result of `Cipher.decrypt` is a binary and not a string:

```
iex(6)> encrypted_value = Repo.get(User, 9).encrypted_api_token
"g7Yd7my6FJOlgRXqCeDDRChoIorKcqXN5qcC0PXei4Ai6VXHYxYFVuevpon%2BIZeQ"
iex(7)> encrypted_value |> Cipher.decrypt
<<87, 77, 111, 49, 68, 55, 99, 99, 81, 97, 57, 117, 116, 76, 104, 98, 84, 119,
  50, 111, 98, 110, 98, 56, 107, 116, 98, 55, 104, 98, 52, 82, 73, 101, 50, 86,
  87, 57, 116, 78, 122, 122, 77, 61, 4, 4, 4, 4>>
iex(8)> encrypted_value |> Cipher.decrypt |> Cipher.encrypt
"g7Yd7my6FJOlgRXqCeDDRChoIorKcqXN5qcC0PXei4Ai6VXHYxYFVuevpon%2BIZeQq9RkvL9%2F1apbFcvzAejXog%3D%3D"
```

Is this intended behavior?

**Comment by rubencaro** [comment]
Hi @johnmosesman ,

No it's not. I can reproduce this behaviour by encrypting with 1.3.1 and decrypting with 1.3.2.

Reviewing actual changes from 1.3.1 to 1.3.2 [here](https://github.com/rubencaro/cipher/compare/1.3.1...1.3.2), I found [the one](https://github.com/rubencaro/cipher/compare/1.3.1...1.3.2#diff-e0d2af272aacc7c631ec03b934c29f5eR258) affecting you. That was introduced to remove Elixir deprecation warnings like this:

`warning: String.rstrip/2 is deprecated, use String.trim_trailing/2 with a binary as second argument`

The mistake was to follow the warning recommendations. Looking at the actual implementation of `String.rstrip/2` [here](https://github.com/elixir-lang/elixir/blob/a311ae403e1f22280ab548bed14f2f8b820d65c2/lib/elixir/lib/string.ex#L629), it is not exactly equivalent to `String.trim_trailing/2` [here](https://github.com/elixir-lang/elixir/blob/a311ae403e1f22280ab548bed14f2f8b820d65c2/lib/elixir/lib/string.ex#L898). Not for `Cipher` low level needs at least, as you proved.

Fixed in https://github.com/rubencaro/cipher/commit/c338e9126c8a1701698e17d295c8dd073c2040c6

**Comment by rubencaro** [comment]
Just released https://hex.pm/packages/cipher/1.3.3.

Nice catch! Thanks!

**Comment by johnmosesman** [comment]
@rubencaro thank you for the quick response.

Using 1.3.3 I'm still getting a different encrypted value

```
iex(8)> enc
"g7Yd7my6FJOlgRXqCeDDRChoIorKcqXN5qcC0PXei4Ai6VXHYxYFVuevpon%2BIZeQ"
iex(9)> enc |> Cipher.decrypt |> Cipher.encrypt
"g7Yd7my6FJOlgRXqCeDDRChoIorKcqXN5qcC0PXei4A1bq4MDcV0UC9QzbciK4uN"
```

**Comment by rubencaro** [comment]
Yes, you will get different encrypted value. That is expected. In general
you should not relay on the actual encrypted value, as it may contain some
noise factor making it always different (at least not always the same).

You should only relay on the decrypted value, which really gives you the
data you wanted to encrypt.

**Comment by johnmosesman** [comment]
@rubencaro Hmm I am surprised to hear that. I had thought the encryption would be reversible both ways.

I can understand the use case for just wanting to store encrypted data, but our case is encrypting API tokens. That way if someone happened to get a copy of our database they wouldn't have everyone's API token. 

To do this we of course need to be able to return decrypted tokens (which works), but then re-encrypt the token and compare it to the original encrypted value. This doesn't work if the encrypted value is different from the one we stored originally.

What causes the change in the encrypted value between versions? As you can see from my first example using 1.3.1 it works there. Are we just using this package in a non-intended way?

**Comment by rubencaro** [comment]
The cause is the change in the 'pad' function, I guess. Similar to what got
fixed in 1.3.3, but for the encrypting part. Before touching that you
should check that anything encrypted with 1.3.2 still gets decrypted. Can
you do that and make a pull request?

**Comment by johnmosesman** [comment]
@rubencaro Thank you for the quick replies. We are in the middle of an imminent product launch for the next couple weeks. I will try to swing back to this after that.
